src/util/http_utils.js

```diff
@@ -664,6 +665,131 @@ function set_amz_headers(req, res) {
+ *
+ * priority is based on:
+ *  - longest matching prefix
+ *  - most matching tags
+ *  - narrowest object size range
+ *
+ * @param {Array<Object>} rules
+ * @param {Object} object_md
+ * @returns {Object|undefined}
+ */
+function get_lifecycle_rule_for_object(rules, object_md) {
+    let matched_rule;
+    let rule_priority = {
+        prefix_len: -1,
+        tag_count: -1,
+        size_span: Infinity,
+    };
+    if (object_md?.key && rules?.length > 0) {
+        for (const rule of rules) {
+            if (rule?.status !== 'Enabled') continue;
+
+            const filter = rule?.filter || {};
+
+            if (filter.prefix && !object_md?.key.startsWith(filter.prefix)) continue;
+
+            if (filter.object_size_greater_than && object_md?.size <= filter.object_size_greater_than) continue;
+            if (filter.object_size_less_than && object_md?.size >= filter.object_size_less_than) continue;
+
+            if (filter.tagging && Array.isArray(filter.tagging)) {
+                const obj_tags = object_md?.tagging || [];
+
+                const matches_all_tags = filter.tagging.every(filter_tag =>
+                    obj_tags.some(obj_tag => obj_tag.key === filter_tag.key && obj_tag.value === filter_tag.value)
+                );
+
+                if (!matches_all_tags) continue;
+            }
+
+            const priority = {
+                prefix_len: (filter?.prefix || '').length,
+                tag_count: Array.isArray(filter?.tagging) ? filter?.tagging.length : 0,
+                size_span: (filter?.object_size_less_than ?? Infinity) - (filter?.object_size_greater_than ?? 0)
+            };
+
+            // compare prefix length
+            const is_more_specific_prefix = priority.prefix_len > rule_priority.prefix_len;
+
+            // compare tag count (if prefixes are equal)
+            const is_more_specific_tags = priority.prefix_len === rule_priority.prefix_len &&
+                                        priority.tag_count > rule_priority.tag_count;
+
+            // compare size span (if prefixes and tags are equal)
+            const is_more_specific_size = priority.prefix_len === rule_priority.prefix_len &&
+                                        priority.tag_count === rule_priority.tag_count &&
+                                        priority.size_span < rule_priority.size_span;
+
+            if (is_more_specific_prefix || is_more_specific_tags || is_more_specific_size) {
+                matched_rule = rule;
+                rule_priority = priority;
+            }
+        }
+    }
+    return matched_rule;
+}
+
+/**
+ * parse_expiration_header converts an expiration rule (either with `date` or `days`)
+ * into an s3 style `x-amz-expiration` header value
+ *
+ * @param {Object} rule
+ * @param {Object} create_time
+ * @returns {string|undefined}
+ *
+ * Example output:
+ *   expiry-date="Thu, 10 Apr 2025 00:00:00 GMT", rule-id="rule_id"
+ */
+function parse_expiration_header(rule, create_time) {
```

Comment:
I don't think its a good name. you don't parse the header, you write it

Reply:
Changed, Thanks